p2p/src/transports/mconnection/mconnection.rs

```diff
@@ -0,0 +1,232 @@
+//! stream.
+//! Spec: https://github.com/tendermint/spec/blob/master/spec/p2p/connection.md#p2p-multiplex-connection
+
+use std::net::{Shutdown, SocketAddr, TcpListener, TcpStream};
+use std::sync::Arc;
+use std::time::Duration;
+
+use ed25519_dalek as ed25519;
+use eyre::{Result, WrapErr};
+
+use super::super::super::secret_connection::{SecretConnection, Version};
+use super::super::super::transport::*;
+
+struct MConnectionTransport {}
+
+struct MConnection {
+    public_key: ed25519::PublicKey,
+    secret_connection: SecretConnection<TcpStream>,
+    stream: TcpStream,
+}
+
+struct MEndpoint {
+    private_key: Arc<ed25519::Keypair>,
+    protocol_version: Version,
+}
+
+struct MIncoming {
+    tcp_listener: TcpListener,
+    private_key: Arc<ed25519::Keypair>,
+    protocol_version: Version,
+}
+
+impl MConnection {
+    /// Opens a TCP connection to a remote host and establishes a secret connection
+    /// [`SecretConnection`].
+    pub fn connect(
+        addr: &SocketAddr,
+        private_key: &ed25519::Keypair,
+        protocol_version: Version,
+    ) -> Result<Self> {
+        let stream = TcpStream::connect(addr)?;
+        let stream_clone = stream.try_clone()?;
+
+        let public_key = private_key.public.clone();
+        let secret_connection = SecretConnection::new(stream, &private_key, protocol_version)?;
+
+        Ok(Self {
+            public_key,
+            secret_connection,
+            stream: stream_clone,
+        })
+    }
+
+    /// Opens a TCP connection to a remote host with a timeout and establishes a secret connection
+    /// [`SecretConnection`].
+    pub fn connect_timeout(
+        addr: &SocketAddr,
+        timeout: Duration,
+        private_key: &ed25519::Keypair,
+        protocol_version: Version,
+    ) -> Result<Self> {
+        let stream = TcpStream::connect_timeout(addr, timeout)?;
+        let stream_clone = stream.try_clone()?;
+
+        let public_key = private_key.public.clone();
+        let secret_connection = SecretConnection::new(stream, &private_key, protocol_version)?;
+
+        Ok(Self {
+            public_key,
+            secret_connection,
+            stream: stream_clone,
+        })
+    }
+}
+
+impl Connection for MConnection {
+    type Error = std::io::Error;
+    type Read = SecretConnection<TcpStream>;
+    type Write = SecretConnection<TcpStream>;
+
+    fn advertised_addrs(&self) -> Vec<SocketAddr> {
+        vec![]
+    }
+
+    fn close(&self) -> Result<()> {
+        self.stream
+            .shutdown(Shutdown::Both)
+            .wrap_err("failed to shutdown")
+    }
+
+    fn local_addr(&self) -> SocketAddr {
+        self.stream.local_addr().unwrap()
+    }
+
+    fn remote_addr(&self) -> SocketAddr {
+        self.stream.peer_addr().unwrap()
+    }
+
+    fn open_bidirectional(
+        &self,
+        _stream_id: &StreamId,
+    ) -> Result<(&Self::Read, &Self::Write), Self::Error> {
+        Ok((&self.secret_connection, &self.secret_connection))
```

Comment:
wonder if I should use smth like Rc or Arc to share secret_connection between multiple callers...